### PR TITLE
virts-621

### PR DESCRIPTION
**Pull request by ArtificialErmine**
Added host filtering mechanism

### CODE REVIEW

File: app/utility/base_planning_svc.py [hunk 1]
```diff
@@ -140,3 +142,21 @@ def _is_fact_bound(fact):
             if not requirement.enforce(combo[0], uf, operation.all_facts()):
                 return False
         return True
+
+    async def _exclude_existing(self, combo, operation):
+        all_hostnames = [agent.paw.split('$')[0].lower() for agent in await self._active_agents(operation)]
```
Comment (wbooth) [inline app/utility/base_planning_svc.py]: Henry is putting in a change that will make this easier and cleaner. See his or out to add host and username as agent attributes

File: app/utility/base_planning_svc.py [hunk 2]
```diff
@@ -44,7 +44,9 @@ class BasePlanningService(BaseService):
                 relevant_facts = await self._build_relevant_facts(variables, operation, agent_facts)
                 valid_facts = await RuleSet(rules=operation.rules).apply_rules(facts=relevant_facts[0])
                 for combo in list(itertools.product(*valid_facts)):
-                    if ability_requirements and not await self._do_enforcements(ability_requirements[link.ability.unique], operation, link, combo):
+                    if ability_requirements and not await \
+                            self._exclude_existing(combo, operation) and not await \
+                            self._do_enforcements(ability_requirements[link.ability.unique], operation, link, combo):
```
Comment (khyberspache) [inline app/utility/base_planning_svc.py]: Looking toward the future... do you foresee this conditional check "growing" with further checks?  I'm working if now would be the right time to move this check condition into it's own private function if we expect further checks.  thoughts?

File: app/utility/base_planning_svc.py [hunk 3]
```diff
@@ -140,3 +142,21 @@ def _is_fact_bound(fact):
...
+        for item in combo:
+            # prevent backwards lateral movement
+            if 'remote.host' in item.trait:
+                if item.value.split('.')[0].lower() in all_hostnames:
+                    return False
+                elif any(h in item.value.split('.')[0].lower() for h in all_hostnames):
+                    return False
+        return True
+
+    async def _active_agents(self, operation):
```
Comment (khyberspache) [inline app/utility/base_planning_svc.py]: I think this function might be useful in the operation object for use elsewhere. Thoughts?

File: app/utility/base_planning_svc.py (hunk 1 above)
Comment (ArtificialErmine) [reply app/utility/base_planning_svc.py]: We've talked about it. If his code gets in before this, I'll update it accordingly.

File: app/utility/base_planning_svc.py (hunk 3 above)
Comment (ArtificialErmine) [reply app/utility/base_planning_svc.py]: Seems reasonable.

File: app/utility/base_planning_svc.py [hunk 4]
```diff
@@ -142,3 +148,14 @@ def _is_fact_bound(fact):
...
+
+    async def _exclude_existing(self, link, operation):
+        all_hostnames = [agent.host for agent in await operation._active_agents()]
+        for item in link.relationships:
+            # prevent backwards lateral movement
+            if 'remote.host' in item.trait:
+                if item.value.split('.')[0].lower() in all_hostnames:
+                    return False
+                elif any(h in item.value.split('.')[0].lower() for h in all_hostnames):
+                    return False
```
Comment (khyberspache) [inline app/utility/base_planning_svc.py]: Shouldn't this be a `if item.value.split('.')[0].lower() in all_hostnames or any(h in item.value.split('.')[0].lower() for h in all_hostnames):`

File: app/objects/c_operation.py [hunk 5]
```diff
@@ -128,6 +128,13 @@ def all_relationships(self):
                 if await self._trust_issues(member):
                     break
 
+    async def _active_agents(self):
+        active = []
```
Comment (privateducky) [inline app/objects/c_operation.py]: move this function down to the private area

Comment (khyberspache) [reply app/objects/c_operation.py]: Ah missed this. Moving now.

**Pull request merged**